core/src/main/scala/cats/data/Continuation.scala

```diff
@@ -0,0 +1,108 @@
+package cats
+package data
+
+import java.io.Serializable
+import scala.annotation.tailrec
+
+sealed abstract class Continuation[O, +I] extends Serializable {
+  def map[I2](fn: I => I2): Continuation[O, I2] =
+    Continuation.Mapped(this, fn)
+
+  def flatMap[I2](fn: I => Continuation[O, I2]): Continuation[O, I2] =
+    Continuation.FlatMapped(this, fn)
+
+  final def apply(fn: I => O): O = {
+    import Continuation._
+    val re = new RunEnv[O]
+    re.eval(this, re.ScalaFn(fn)).get
+  }
+}
+
+object Continuation {
+  def pure[O] = new PureBuilder[O]
+  class PureBuilder[O] {
+    def apply[I](i: I): Continuation[O, I] = Const(i)
+  }
+  def from[I, O](fn: (I => O) => O): Continuation[O, I] = Cont(fn)
+  def from2[I1, I2, O](call: (Function2[I1, I2, O]) => O): Continuation[O, (I1, I2)] = from { fn: (((I1, I2)) => O) =>
+    call { (i1: I1, i2: I2) => fn((i1, i2)) }
+  }
+
+  implicit def catsDataContinuationMonad[O]: Monad[Continuation[O, ?]] = new Monad[Continuation[O, ?]] {
+    def pure[I](i: I): Continuation[O, I] = Const(i)
+    override def map[I1, I2](f: Continuation[O, I1])(fn: I1 => I2): Continuation[O, I2] = f.map(fn)
+    def flatMap[I1, I2](f: Continuation[O, I1])(fn1: I1 => Continuation[O, I2]): Continuation[O, I2] = f.flatMap(fn1)
+    /**
+     * flatMap is trampolined, BUT the stack depth grows proportional to the number of `from` calls
+     */
+    def tailRecM[A, B](a: A)(fn: A => Continuation[O, Either[A, B]]): Continuation[O, B] =
+      fn(a).flatMap {
+        case Right(b) => Const(b)
+        case Left(a) => tailRecM(a)(fn)
+      }
+  }
+
+  private case class Const[O, I](i: I) extends Continuation[O, I]
+
```

Comment:
Don't know if also marking these as `final` does anything but maybe for style consistency?